src/main/java/com/snowflake/kafka/connector/internal/SnowflakeInternalStage.java

```diff
@@ -65,72 +106,139 @@ public SnowflakeInternalStage(
-        // If the backend is not GCP, we cache the credential. Otherwise throw error.
-        // transfer metadata list must only have one element
-        SnowflakeFileTransferMetadataV1 fileTransferMetadata =
-            (SnowflakeFileTransferMetadataV1) agent.getFileTransferMetadatas().get(0);
-        if (fileTransferMetadata.getStageInfo().getStageType() != StageInfo.StageType.GCS) {
-          // Overwrite the credential to be used
-          credential =
-              new SnowflakeMetadataWithExpiration(fileTransferMetadata, System.currentTimeMillis());
-          storageInfoCache.put(stage, credential);
-        } else {
-          throw SnowflakeErrors.ERROR_5017.getException();
-        }
-      }
+      SnowflakeMetadataWithExpiration credential = storageInfoCache.getOrDefault(stageName, null);
 
-      SnowflakeFileTransferMetadataV1 fileTransferMetadata = credential.fileTransferMetadata;
-      // Set filename to be uploaded
-      fileTransferMetadata.setPresignedUrlFileName(fullFilePath);
-
-      byte[] dataBytes = data.getBytes(StandardCharsets.UTF_8);
-      InputStream inStream = new ByteArrayInputStream(dataBytes);
-
-      // This uploadWithoutConnection api cannot handle expired credentials very well.
-      // Need to prevent passing expired credential to it.
-      try {
-        SnowflakeFileTransferAgent.uploadWithoutConnection(
-            SnowflakeFileTransferConfig.Builder.newInstance()
-                .setSnowflakeFileTransferMetadata(fileTransferMetadata)
-                .setUploadStream(inStream)
-                .setRequireCompress(true)
-                .setOcspMode(OCSPMode.FAIL_OPEN)
-                .setProxyProperties(proxyProperties)
-                .build());
-      } catch (Throwable t) {
-        // If this api encounters error, invalid the cached credentials
-        // Caller will retry this function
-        logWarn("uploadWithoutConnection encountered an error for fileName:{}", fullFilePath);
-        storageInfoCache.remove(stage);
-        throw t;
+      if (!isCredentialValid(credential, stageType)) {
+        logDebug(
+            "Query credential(Refreshing Credentials) for stageName:{}, filePath:{}",
+            stageName,
+            fullFilePath);
+        refreshCredentials(stageName, stageType, fullFilePath);
       }
+    } catch (Exception e) {
+      logWarn(
+          "Failed to refresh Credentials for stageName:{}, filePath:{}", stageName, fullFilePath);
+      throw SnowflakeErrors.ERROR_5018.getException(e.getMessage());
+    }
+
+    // Lets fetch the file transfer metadata from cache again. (Because we might have refreshed
+    // them)
+    SnowflakeFileTransferMetadataV1 fileTransferMetadata =
+        storageInfoCache.get(stageName).fileTransferMetadata;
+    // Set filename to be uploaded
+    // This set is not useful in GCS since there is a bug in JDBC which doesnt use destFileName.
+    fileTransferMetadata.setPresignedUrlFileName(fullFilePath);
```

Comment:
Is there a JIRA to track the JDBC bug? Worth adding that in code

Reply:
We don't have a jira but have a bug assigned (Linked in Description) Will create a JIRA.